NAPS2.Lib.Tests/Dependencies/DownloadControllerTests.cs

```diff
@@ -0,0 +1,189 @@
+﻿using NAPS2.Dependencies;
+using NAPS2.Sdk.Tests;
+using RichardSzalay.MockHttp;
+using Xunit;
+
+namespace NAPS2.Lib.Tests.Dependencies;
+
+public class DownloadControllerTests : ContextualTests
+{
+    private const string AnimalsSHA1 = "1739ff74f5d26f6c42d7a3fc438d615a93d38bd4";
+    private const string StockDogJpegSHA1 = "ca1f89964bdb345e97804784eefa49a00dbf7207";
+
+    private class EventHandlerWrapper
+    {
+        private readonly Dictionary<EventHandler<EventArgs>, EventHandler> handlerMapping = new();
+
+        public EventHandler Attach(EventHandler<EventArgs> handler)
+        {
+            Assert.DoesNotContain(handler, handlerMapping.Keys);
+            handlerMapping[handler] = (s, e) => handler(s, e);
+            return handlerMapping[handler];
+        }
+
+        public EventHandler Detach(EventHandler<EventArgs> handler)
+        {
+            EventHandler handler2 = handlerMapping[handler];
+            handlerMapping.Remove(handler);
+            return handler2;
+        }
+    }
+
+    private class FailHandler : IDownloadHandler
+    {
+        public DownloadInfo DownloadInfo { get; }
+
+        public FailHandler(List<DownloadMirror> downloadMirrors, string sha1 = StockDogJpegSHA1)
+        {
+            DownloadInfo = new("temp.gz", downloadMirrors, 0, sha1, DownloadFormat.Gzip);
+        }
+
+        public void Install(string sourcePath)
+        {
+            Assert.Fail("The file callback should not have been called.");
```

Comment:
Assertions in this method (and in GzipHandler.Install) don't actually work as the exception is caught in DownloadController.

Rather than having the IDownloadHandler interface with these custom implementations, I would recommend using a `Mock<IExternalComponent>`. Then you either `.Verify` that the method is called or `.VerifyNoOtherCalls` to confirm it isn't called.